### PR TITLE
Azure driver: Fix issue #4244

### LINKED ISSUE #4244: Azure: Associated subnet in routing table is deleted during "Configuring subnet"
If you start a docker-machine in Microsoft Azure cloud and if you use a subnet that is precreated to use it in a Azure routing table, the subnet entry gets deleted. This setup is used in environments where the company has a site-to-site connection to the Azure cloud for private on-premise usage.

# How to test

You have to add a Azure-RT entry beforehand.

## Start docker-machine:
```
(runner-bb9e4941-auto-scale-runners-1504681201-c7e15b19) Configuring subnet.  name="SUBNET-LAN-HostingWeb-WE" vnet="VNET-WE" cidr="10.191.2.0/24"  driver=azure name=runner-bb9e4941-auto-scale-runners-1504681201-c7e15b19 operation=create
WARNING: Failed to update executor docker+machine for bb9e4941 No free machines that can process builds                                                                                                                                      
WARNING: Failed to update executor docker+machine for bb9e4941 No free machines that can process builds
(runner-bb9e4941-auto-scale-runners-1504681201-c7e15b19) Not creating a public IP address.  driver=azure name=runner-bb9e4941-auto-scale-runners-1504681201-c7e15b19 operation=create
(runner-bb9e4941-auto-scale-runners-1504681201-c7e15b19) Creating network interface.  name="runner-bb9e4941-auto-scale-runners-1504681201-c7e15b19-nic"  driver=azure name=runner-bb9e4941-auto-scale-runners-1504681201-c7e15b19 operation=c
reate                                                                                                  
(runner-bb9e4941-auto-scale-runners-1504681201-c7e15b19) Creating storage account.  name="vhdsl4y2rwfs3eh8s8ib1bxh" location="West Europe" sku=Standard_LRS  driver=azure name=runner-bb9e4941-auto-scale-runners-1504681201-c7e15b19 operati
on=create
```

Immediately after the line "Configuring subnet." the associated routing table subnet entry is deleted.

## Debug run of docker-machine

This is the debug run of the native docker-machine create command.

### version
```
docker-machine version 0.12.2, build 9371605
```

### command
```
docker-machine --debug create -d azure \
	--azure-subscription-id ssssad \
	--azure-location "ssss" \
	--azure-ssh-user gitlab \
	--azure-resource-group aaa \
	--azure-vnet xxx \
	--azure-subnet yyy \
	--azure-subnet-prefix 10.191.2.0/24 \
	--azure-use-private-ip \
	--azure-no-public-ip \
	dockervm-gitlab1
```

### result
```
(dockervm-gitlab1) Configuring subnet.  name="SUBNET-LAN-HostingWeb-WE" vnet="VNET-WE" cidr="10.191.2.0/24"
(dockervm-gitlab1) DBG | Azure request  method="PUT" request=
(dockervm-gitlab1) DBG | Azure response  status="200 OK" method="PUT" request=
(dockervm-gitlab1) DBG | Azure request  method="GET" request=
(dockervm-gitlab1) DBG | Azure response  status="200 OK" method="GET" request=
(dockervm-gitlab1) DBG | Azure response  status="200 OK" method="GET" request=
(dockervm-gitlab1) DBG | Azure request  request=
(dockervm-gitlab1) DBG | Azure response  request= status="200 OK" method="GET"
(dockervm-gitlab1) DBG | Azure response  status="200 OK" method="GET" request=
(dockervm-gitlab1) Not creating a public IP address.
```

## Conclusion

During the createorupdate API method [documented here](https://docs.microsoft.com/en-us/rest/api/virtualnetwork/subnets/createorupdate), the routing table entry is deleted. The API call responses with code 200 so according to the API documentation the subnet is updated successful but the routing table entry is destroyed.

We have to specify the subnet during the docker-machine run because it defaults to an 192.* CIDR. That is not useful in our case because we have on-premise routing to Microsoft Azure.

## Possible fix

I will link this issue to the Microsoft Azure go SDK issue tracker on Github too. If this behavior is a result of the API method createorupdate, this should be fixed there maybe. If this is an expected behavior there, then this should be fixed here. We have drilled down the called API resource and are pretty sure that just createorupdate is used and nothing more.

In the mean time it will be better to get the existing subnets and only call create or update if the subnet does not already exist.

### PR DESCRIPTION
**Pull request by kleinsasserm**
This will fix my issue #4244. We are currently using this for our private Gitlab environment to use the auto scaling feature from Gitlab runner with docker-machine and Microsoft Azure.

Signed-off-by: Mario Kleinsasser <mario.kleinsasser@gmail.com>

### CODE REVIEW

**Comment by kleinsasserm** [comment]
@shin Would it be possible for you to check why the travis build failed please? Because the error I in travis is not clear to me and the build of another patch of mine worked without problems. Thank you!

**Comment by kleinsasserm** [comment]
@shin @StefanScherer @nathanleclaire any comments on this from the committers? Is something missing or unclear? Can I do something to get this commited? Thank you!

**Comment by nathanleclaire** [comment]
ping @shin-

**Comment by kleinsasserm** [comment]
Oh, now I see my mistake 😊 ... shin- not shin. 😄 Sorry!

**Comment by shin-** [comment]
Hi @kleinsasserm , I believe you need to run a `gofmt` pass on this PR for the Travis tests to pass.

**Comment by kleinsasserm** [comment]
@shin- I wrote you an email offline, because I do not spam the PR. I do not get why the Travis build is complaining about lint, but ```make lint``` doesn't. Maybe you can have a look in the next days.

**Comment by kleinsasserm** [comment]
@shin- Can you have a look at my email? I would like to do this PR the correct way, but I need some help here. Anyone else who can have a look at my PR and my mistake(s) 😊 ?

**Comment by shin-** [comment]
This is what I'm seeing on Travis in terms of linting errors:
```
drivers/azure/azure.go:393:2: redundant if ...; err != nil check, just return error instead.
drivers/azure/azure.go:436:2: redundant if ...; err != nil check, just return error instead.
drivers/exoscale/exoscale.go:458:2: redundant if ...; err != nil check, just return error instead.
drivers/hyperv/hyperv.go:180:2: redundant if ...; err != nil check, just return error instead.
drivers/vmwarevcloudair/vcloudair.go:448:2: redundant if ...; err != nil check, just return error instead.
drivers/vmwarevsphere/vsphere.go:938:2: redundant if ...; err != nil check, just return error instead.
libmachine/persist/filestore.go:56:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/arch.go:150:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/boot2docker.go:258:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/coreos.go:130:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/debian.go:141:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/rancheros.go:139:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/redhat.go:113:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/redhat.go:181:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/suse.go:205:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/ubuntu_systemd.go:151:2: redundant if ...; err != nil check, just return error instead.
libmachine/provision/ubuntu_upstart.go:160:2: redundant if ...; err != nil check, just return error instead.
make: *** [lint] Error 1
mk/validate.mk:18: recipe for target 'lint' failed
make: *** [build validate] Error 2
```

None of those seem related to your changes, so I'm guessing they might be caused by an update of the linter that made it stricter. I'll look into fixing those in master when I have some time. Thank you for your patience.

**Comment by kleinsasserm** [comment]
You are welcome! I just wanted to make sure that it wasn't my mistake :smile:

**Comment by shin-** [comment]
@kleinsasserm I just merged #4266 , so if you can rebase this PR we should be able to move forward. Thank you for your patience! :+1:

File: drivers/azure/azureutil/azureutil.go [hunk 1]
```diff
@@ -232,21 +232,31 @@ func (a AzureClient) GetSubnet(resourceGroup, virtualNetwork, name string) (netw
 	return a.subnetsClient().Get(resourceGroup, virtualNetwork, name, "")
 }
 
+// CreateSubnet creates or updates a subnet if it does not already exist.
 func (a AzureClient) CreateSubnet(ctx *DeploymentContext, resourceGroup, virtualNetwork, name, subnetPrefix string) error {
+	subnet, err := a.subnetsClient().Get(resourceGroup, virtualNetwork, name, "")
+	if subnet.Name != nil {
```
Comment (shin-) [inline drivers/azure/azureutil/azureutil.go]: I would rather handle this using the `err` value. 
- If `err` is nil, that means the subnet exists and we can exit. 
- If it does not, we should cast it to the `autorest.DetailedError` type. 
  - If `err.StatusCode` is `http.NotFound`, we can proceed with creating the subnet; 
  - if it's something else, we should probably log and abort.

Comment (kleinsasserm) [reply drivers/azure/azureutil/azureutil.go]: Hm, I thought that ```err``` is only providing information but connection problems and so on, but well I will have a look,

**Comment by kleinsasserm** [comment]
@shin- I've updated the PR. Hopefully it fits the needs and I made it the correct way. I think that this version is more informative too but it "feels" a little bit complicated. Should we log the ```err``` at this point as ```log.Warn```? I looked at the code of the other methods but I am not sure. Sorry for the delay, but I was busy in the last days and it took me some time to test this change and to understand the returns in detail. :smiley:

**Comment by shin-** [comment]
Looks pretty good to me! Thanks for following up, and thank you again for your contribution!

**Comment by kleinsasserm** [comment]
You are welcome! Thank you!

**Pull request merged**